### ISSUE TITLE
compiling boost_python

### ISSUE THREAD

**Issue by gkroone** [SynoCommunity/spksrc#1648]
I'm new to programming and spksrc, but giving it a try anyhow, but I'm having trouble compiling latest revisions of domoticz due to it requiring boost_python.

in the configure part of the MakeFile for boost, it creates a file called: 
work-x86-5.1/boost_1_53_0/tools/build/v2/user-config.jam

when I modify the boost Makefile to list:

    @echo "using gcc : `$(TC_PATH)$(TC_PREFIX)gcc -dumpversion` : $(TC_PATH)$(TC_PREFIX)g++ : <address-model>\"$(ADDRESS_MODEL)\" <cflags>\"$(CFLAGS)\" <cxxflags>\"$(CXXFLAGS)\" <linkflags>\" -L$(STAGING_INSTALL_PREFIX)/lib/$(PYTHON_INC_DIR) $(LDFLAGS)\" <link>\"shared\" ; " > $(WORK_DIR)/$(PKG_DIR)/tools/build/v2/user-config.jam

to basically add into 
    \<linkflags\>
    -L$(STAGING_INSTALL_PREFIX)/lib/$(PYTHON_INC_DIR)

and also in the Domoticz Makefile, I add 
    PYTHON_INC_DIR="python2.7" 

then the boost compiles correctly, but I was unable to figure out how to make it compile by just modifying the Domoticz MakeFile.
I tried many things.  adding e.g. 
    ADDITIONAL_LDFLAGS = "-L$(STAGING_INSTALL_PREFIX)/lib/$(PYTHON_INC_DIR)"
did not work.

how should I do this properly? 

Is this a problem with boost MakeFile or with Domoticz MakeFile?

**Comment by gkroone** [comment]
@LouiS22 I have committed my change to the boost Makefile in:
https://github.com/gkroone/spksrc/commits/openzwave/
 I have not yet done a full compile on this yet(compile is still running overnight)

**Comment by LouiS22** [comment]
Unfortunately still a no go

Command "/home/attila/spksrc/native/python/work-native/install/usr/local/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-vMf3kP/cffi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-G_h4uW-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-vMf3kP/cffi
make[5]: *** [myPostInstall] Error 1
make[5]: Leaving directory `/home/attila/spksrc/native/python'
make[4]: *** [depend_target] Error 2
make[4]: Leaving directory `/home/attila/spksrc/cross/python'
make[3]: *** [depend_target] Error 2
make[3]: Leaving directory `/home/attila/spksrc/cross/boost'
make[2]: *** [depend_target] Error 2
make[2]: Leaving directory `/home/attila/spksrc/cross/domoticz'
make[1]: *** [depend_target] Error 2
make[1]: Leaving directory `/home/attila/spksrc/spk/domoticz'
make: [arch-88f6281] Error 2 (ignored)
attila@debian:~/spksrc/spk/domoticz$

**Comment by gkroone** [comment]
Try to compile native/python first.

**Comment by LouiS22** [comment]
Tried, with no luck. Even the native python compile finished without errors, running yours makes the failed to compile error message.

**Comment by gkroone** [comment]
I am now also seeing problems in the build again too, but they're different from yours. I made several changes at once, as I undid some stuff in the domoticz Makefile which I though might not be needed after all and wanted to verify that.

here's what happened for me during the first build after a fresh pull of the git repo:

    Complete output from command /home/geert/git/spksrc/native/python/work-native/install/usr/local/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-ptvQhk/cffi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-NIpKU6-record/install-record.txt --single-version-externally-managed --compile:
    Package libffi was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libffi.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libffi' found

that error was fixed by compiling native python.

another make arch-x86 for domoticz then ended up in the boost_python:
common.mkdir bin.v2/libs/python
common.mkdir bin.v2/libs/python/build
common.mkdir bin.v2/libs/python/build/gcc-4.7.3
common.mkdir bin.v2/libs/python/build/gcc-4.7.3/release
common.mkdir bin.v2/libs/python/build/gcc-4.7.3/release/threading-multi
gcc.compile.c++ bin.v2/libs/python/build/gcc-4.7.3/release/threading-multi/numeric.o
In file included from ./boost/python/detail/prefix.hpp:13:0,
                 from ./boost/python/numeric.hpp:8,
                 from libs/python/src/numeric.cpp:6:
./boost/python/detail/wrap_python.hpp:50:23: fatal error: pyconfig.h: No such file or directory
compilation terminated.

I thought I had fixed this, as the previous time I managed to succesfully compiled domoticz r2382, i eventually got through this by manually editting the file 
    work-x86-5.1/boost_1_53_0/tools/build/v2/user-config.jam
by changing the \<linkflags\> to point to the directory where  pyconfig.h and other .h files are placed. In my case, that is /home/geert/git/spksrc/spk/domoticz/work-x86-5.1/install/usr/local/domoticz/lib/python2.7

however in this case it looks like that's not enough. 
I think I might need to revert my last change to the cross/domoticz/Makefile and re-do the boost configuration.

I'll be doing this after I reverted the beforementionned Makefile:
rm -rf work-x86-5.1/boost*
rm work-x86-5.1/.boost
make arch-x86

I am not going to commit any more changes until I am sure the whole thing compiles fine for me after a make clean. this will take another day to fully compile.

If you have any luck in the meantime, let me know.

**Comment by LouiS22** [comment]
Hm, I'm trying to compile to marvel 88f6281, maybe that's why the different error messages. What I noticed: even if I build native python first (which compiles without errors), I'm not able to compile it after - no matter what I'm getting the compilation error messages.

**Comment by gkroone** [comment]
After adding back the additional env stuff into cross/domoticz/Makefile:
    -ENV += BOOST_LIBRARIES="$(BOOST_LIBRARIES)" CMAKE_INCLUDE_PATH="$(STAGING_INSTALL_PREFIX)/include" CMAKE_LIBRARY_PATH="$(STAGING_INSTALL_PREFIX)/lib" PCH_COMPILE_FLAGS="$(CFLAGS)" PYTHON_VERSION="2.7.8" PYTHON_INC_DIR=python2.7/
    +ENV += BOOST_LIBRARIES="$(BOOST_LIBRARIES)" CMAKE_INCLUDE_PATH="$(STAGING_INSTALL_PREFIX)/include" CMAKE_LIBRARY_PATH="$(STAGING_INSTALL_PREFIX)/lib" PCH_COMPILE_FLAGS="$(CFLAGS)" PYTHON_VERSION="2.7" PYTHON_INC_DIR="python2.7" PYTHON_CPPFLAGS="-I$(STAGING_INSTALL_PREFIX)/lib/$(PYTHON_INC_DIR)" ADDITIONAL_CFLAGS="-I$(STAGING_INSTALL_PREFIX)/include/python2.7" ADDITIONAL_LDFLAGS="-L$(STAGING_INSTALL_PREFIX)/lib/python2.7"


And restarting the boost stuff, it just compiled again for me.

I will now do a make clean, and start over from the start again with this.

**Comment by gkroone** [comment]
I don't fully understand what's going on, but last night's build failed again on the same error message related to missing pyconfig.h

And when changing no makefile at all, just removing the boost directory and the  .boost files, and restarting the build, it works fine. This may explain why i initially keep seeing success after every change i make to a makefile

**Comment by gkroone** [comment]
@LouiS22 I've tried to build for marvel 88f6281 and faced the same issue. initial build fails, but then all I do is the following:
mv work-88f6281-5.1/boost_1_53_0 work-88f6281-5.1/boost_1_53_0.bak
rm work-88f6281-5.1/.boost-configure_done
rm work-88f6281-5./.boost-extract_done
rm work-88f6281-5.1/.boost-patch_done
make arch-88f6281

and the compile completes just fine.

I found this article: #516 describing the exact problem.

I uncommented the line in cross/boost/Makefile
    -include $(WORK_DIR)/python-cc.mk

and added to the DEPENDS in domoticz makefile: /cross/python
also removed  my other changes in the makefile for boost and domoticz that are related to boost_python compilation.

now it compiles just fine without error in the boost compilation.
